src/main/java/com/github/manolo8/darkbot/gui/zones/ZoneEditor.java

```diff
@@ -85,60 +66,52 @@ public void mouseMoved(MouseEvent e) {
 
     void setup(Main main, java.util.Map<Integer, ZoneInfo> zonesByMap) {
         super.setup(main);
-        if (mapChange != null) main.mapManager.mapChange.remove(mapChange);
-        main.mapManager.mapChange.add(mapChange = map -> {
-            zoneInfo = zonesByMap.computeIfAbsent(map.id, id -> new ZoneInfo(config.BOT_SETTINGS.OTHER.ZONE_RESOLUTION));
-            ZoneEditor.this.repaint();
-        });
-        zoneInfo = zonesByMap.get(main.hero.map.id);
+
+        main.pluginAPI.requireAPI(EventBrokerAPI.class).registerListener(this);
+
+        this.infosDrawer = main.pluginAPI.requireInstance(InfosDrawer.class);
+        this.zonesDrawer = main.pluginAPI.requireInstance(ZonesDrawer.class);
+        this.constantEntitiesDrawer = main.pluginAPI.requireInstance(ConstantEntitiesDrawer.class);
+
+        this.zonesByMap = zonesByMap;
+        this.zoneInfo = zonesByMap.get(main.hero.map.id);
     }
 
     @Override
-    protected void paintComponent(Graphics g) {
-        Graphics2D g2 = setupDraw(g);
+    protected void onPaint() {
+        zonesDrawer.drawZones(mapGraphics);
 
-        synchronized (Main.UPDATE_LOCKER) {
-            drawZones(g2);
-            drawStaticEntities(g2);
-            drawMap(g2);
-        }
+        constantEntitiesDrawer.drawPortals(mapGraphics);
+        constantEntitiesDrawer.drawBattleStations(mapGraphics);
+        constantEntitiesDrawer.drawStations(mapGraphics);
```

Comment:
```suggestion
        constantEntitiesDrawer.onDraw(mapGraphics);
```

You can draw all constant entities, it is equivalent to this since zones no longer are inside there (and they never should've been)